hypershift-operator/controllers/nodepool/kubevirt/imagecacher.go

```diff
@@ -0,0 +1,247 @@
+
+	if nodePool.Spec.Platform.Kubevirt == nil {
+		// should never happen; but since CacheImage is exposed, we need to protect it from wrong inputs.
+		return fmt.Errorf("nodePool does not contain KubeVirt configurations")
+	}
+
+	dvList, err := getCacheDVs(ctx, cl, nodePool.Spec.ClusterName)
+	if err != nil {
+		return err
+	}
+
+	oldDVs := make([]v1beta1.DataVolume, 0)
+	var dvName string
+	for _, dv := range dvList {
+		if (len(dvName) == 0) && (dv.Annotations[bootImageDVAnnotationHash] == qi.hash) {
+			dvName = dv.Name
+		} else {
+			oldDVs = append(oldDVs, dv)
+		}
+	}
+
+	// if no DV with the required hash was found
+	if len(dvName) == 0 {
+		logger.Info("couldn't find boot image cache DataVolume; creating it...")
+		dv, err := qi.createDVForCache(ctx, cl, nodePool)
+		if err != nil {
+			return err
+		}
+		dvName = dv.Name
+	}
+
+	if len(dvName) == 0 {
+		return fmt.Errorf("can't get the name of the boot image cache DataVolume")
+	}
+	qi.dvName = dvName
+	qi.cleanOldCaches(ctx, cl, oldDVs)
+
+	return nil
+}
+
+func (qi *cachedQCOWImage) cleanOldCaches(ctx context.Context, cl client.Client, oldDVs []v1beta1.DataVolume) {
+	logger := ctrl.LoggerFrom(ctx)
+	for _, oldDV := range oldDVs {
+		if oldDV.DeletionTimestamp == nil {
+			logger.Info("deleting an old boot image cache DataVolume", "namespace", oldDV.Namespace, "DataVolume name", oldDV.Name)
+			err := cl.Delete(ctx, &oldDV)
+			if err != nil {
+				logger.Error(err, fmt.Sprintf("failed to delete an old DataVolume; namespace: %s, name: %s", oldDV.Namespace, oldDV.Name))
+			}
+		}
+	}
+}
+
+func (qi *cachedQCOWImage) createDVForCache(ctx context.Context, cl client.Client, nodePool *hyperv1.NodePool) (*v1beta1.DataVolume, error) {
+	dv := qi.buildDVForCache(nodePool)
+
+	err := cl.Create(ctx, dv)
+	if err != nil && !errors.IsAlreadyExists(err) {
+		return nil, fmt.Errorf("failed to create a DataVolume for the boot image cache ; %w", err)
+	}
+
+	return dv, nil
+}
+
+func (qi *cachedQCOWImage) getDVSourceForVMTemplate() *v1beta1.DataVolumeSource {
+	return &v1beta1.DataVolumeSource{
+		PVC: &v1beta1.DataVolumeSourcePVC{
+			Namespace: qi.namespace,
+			Name:      qi.dvName,
+		},
+	}
+}
+
+func (qi *cachedQCOWImage) buildDVForCache(nodePool *hyperv1.NodePool) *v1beta1.DataVolume {
+	var (
+		volumeSize   apiresource.Quantity
+		storageClass *string
+	)
+
+	kvPlatform := nodePool.Spec.Platform.Kubevirt
+
+	if kvPlatform.RootVolume != nil && kvPlatform.RootVolume.Persistent != nil {
+		if kvPlatform.RootVolume.Persistent.Size != nil {
+			volumeSize = *kvPlatform.RootVolume.Persistent.Size
+		}
+		storageClass = kvPlatform.RootVolume.Persistent.StorageClass
+	}
+
+	return &v1beta1.DataVolume{
+		ObjectMeta: metav1.ObjectMeta{
+			GenerateName: bootImageNamePrefix,
+			Namespace:    qi.namespace,
+			Labels: map[string]string{
+				bootImageDVLabelRoleName:         bootImageDVLabelRoleValue,
+				bootImageDVLabelHosedClusterName: nodePool.Spec.ClusterName,
+			},
```

Comment:
I think we need to add the hosted cluster infra id here. similar to what I'm doing here, https://github.com/openshift/hypershift/pull/2092